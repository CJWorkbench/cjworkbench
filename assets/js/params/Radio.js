// Simple wrapper over HTML <select>
import React from 'react'
import PropTypes from 'prop-types'

export default class RadioParam extends React.PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired, // <input name=...>
    fieldId: PropTypes.string.isRequired, // <input id=...>
    items: PropTypes.string,  // like 'Apple|Banana|Kitten' ... DEPRECATED: prefer `enumOptions`
    enumOptions: PropTypes.arrayOf(PropTypes.shape({
      value: PropTypes.any.isRequired,
      label: PropTypes.string.isRequired,
    }).isRequired), // should be .isRequired).isRequired, but for now we accept .items instead
    value: PropTypes.any.isRequired,
    isReadOnly: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired // onChange(newIndex) => undefined
  }

  onChange = (ev) => {
    // <input value=...> is always a String. Our <option> values aren't.
    // Find the option matching the string, and return it.
    const strValue = ev.target.value
    const option = this.enumOptions.find(({ value }) => String(value) === strValue)
    this.props.onChange(option.value)
  }

  get enumOptions () {
    const { enumOptions, items } = this.props
    if (enumOptions) return enumOptions

    // Handle deprecated `this.props.items`
    return items.split('|').map((label, value) => ({ label, value })) // `value` is int array index
  }

  render() {
    const { items, name, fieldId, isReadOnly, value } = this.props
    const selectedValue = value
    const optionComponents = this.enumOptions.map(({ value, label }, i) => (
      <label key={i} className='t-d-gray content-1'>
        <input
          type='radio'
          name={`${name}`}
          id={`${fieldId}${i === 0 ? '' : ('_' + i)}`}
          className='radio-button'
          value={String(value)}
          checked={String(value) === String(selectedValue)}
          onChange={this.onChange}
          disabled={this.props.isReadOnly}
        />
        <span className="button"></span>
        {label}
      </label>
    ))

    return (
      <div className='button-group'>
        {optionComponents}
      </div>
    )
  }
}
