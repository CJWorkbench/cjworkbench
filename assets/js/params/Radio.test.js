import React from 'react'
import { mount } from 'enzyme'

import Radio from './Radio'

describe('Radio', () => {
  const wrapper = (extraProps = {}) => mount(
    <Radio
      isReadOnly={false}
      name='radio'
      fieldId='radio'
      value={0}
      onChange={jest.fn()}
      {...extraProps}
    />
  )

  it('renders correctly', () => {
    const w = wrapper({ enumOptions: [ { value: 'x', label: 'X' } ] })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders number of buttons correctly', () => {
    const w = wrapper({ enumOptions: [
      { value: 'x', label: 'X' },
      { value: 'y', label: 'Y' },
      { value: 'z', label: 'Z' }
    ]})
    expect(w.find('input[type="radio"]')).toHaveLength(3)
  })

  it('render and handle click of deprecated "items" (as opposed to "enumOptions")', () => {
    const w = wrapper({ items: 'first|second' })
    w.find('input[value="1"]').simulate('change', { target: { value: '1' } })
    expect(w.prop('onChange')).toHaveBeenCalledWith(1)
  })

  it('render and handle click of non-String values', () => {
    const w = wrapper({
      enumOptions: [
        { value: 0, label: 'first' },
        { value: 1, label: 'second' }
      ]
    })
    w.find('input[value="1"]').simulate('change', { target: { value: '1' } })
    expect(w.prop('onChange')).toHaveBeenCalledWith(1)
  })

  it('should be disabled when read only', () => {
    // HTML5 "readOnly" input is editable. Oops.
    const w = wrapper({
      isReadOnly: true,
      name: 'foo',
      enumOptions: [
        { value: 'x', label: 'off' },
        { value: 'y', label: 'on' },
      ]
    })
    expect(w.find('input[value="x"]').prop('disabled')).toBe(true)
    expect(w.find('input[value="y"]').prop('disabled')).toBe(true)
  })
})
