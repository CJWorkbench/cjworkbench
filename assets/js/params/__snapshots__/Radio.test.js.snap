// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<RadioParam
  enumOptions={
    Array [
      Object {
        "label": "X",
        "value": "x",
      },
    ]
  }
  fieldId="radio"
  isReadOnly={false}
  name="radio"
  onChange={[MockFunction]}
  value={0}
>
  <div
    className="radio-options"
  >
    <label
      className="t-d-gray content-1"
      key="0"
    >
      <input
        checked={false}
        className="radio-button"
        disabled={false}
        id="radio"
        name="radio"
        onChange={[Function]}
        type="radio"
        value="x"
      />
      <span
        className="button"
      />
      X
    </label>
  </div>
</RadioParam>
`;
